examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+      dsp.block(1,0).reinit (dof_p, dof_u);
+      dsp.block(0,1).reinit (dof_u, dof_p);
+      dsp.block(1,1).reinit (dof_p, dof_p);
+      dsp.collect_sizes();
+
+      DoFTools::make_sparsity_pattern (dof_handler, dsp, nonzero_constraints);
+      sparsity_pattern.copy_from (dsp);
+    }
+
+    system_matrix.reinit (sparsity_pattern);
+
+
+    present_solution.reinit (2);
+    present_solution.block(0).reinit (dof_u);
+    present_solution.block(1).reinit (dof_p);
+    present_solution.collect_sizes ();
+
+    newton_update.reinit (2);
+    newton_update.block(0).reinit (dof_u);
+    newton_update.block(1).reinit (dof_p);
+    newton_update.collect_sizes ();
+
+    system_rhs.reinit (2);
+    system_rhs.block(0).reinit (dof_u);
+    system_rhs.block(1).reinit (dof_p);
+    system_rhs.collect_sizes ();
+  }
+
+  // @sect4{Navier_Stokes_Newton::assemble_NavierStokes_system}
+  // This function builds the system matrix and right hand side that we actually work on. We can see the function contains
+  // four arguments. "initial_step" is set up for applying different constraints(nonzero for the initial step and zero for
+  // the others). "left" and "right" are flags to determine whether to assemble system matrix and right hand side vector,
+  // respectively. The last argument "alpha" is used while executing line search: it acts as the tentative weight of
+  // Newton's update term.
+
+  template <int dim>
+  void Navier_Stokes_Newton<dim>::assemble_NavierStokes_system(bool initial_step,
+      bool assemble_matrix,
+      bool assemble_rhs)
+  {
+    if (assemble_matrix)
+      {
+        system_matrix    = 0;
+      }
+
+    if (assemble_rhs)
+      {
+        system_rhs       = 0;
+      }
+
+
+
+    QGauss<dim>   quadrature_formula(degree+2);
+
+    FEValues<dim> fe_values (fe,
+                             quadrature_formula,
+                             update_values |
+                             update_quadrature_points |
+                             update_JxW_values |
+                             update_gradients );
+
+    const unsigned int   dofs_per_cell = fe.dofs_per_cell;
+    const unsigned int   n_q_points    = quadrature_formula.size();
+
+    const FEValuesExtractors::Vector velocities (0);
+    const FEValuesExtractors::Scalar pressure (dim);
+
+    FullMatrix<double>   local_matrix (dofs_per_cell, dofs_per_cell);
+    Vector<double>       local_rhs    (dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    // For the linearized system, we create temporary storages for present velocity and its divergence and gradient, and
+    // present pressure. In practice, they are all obtained through their sharp functions at quadrature points. This work
+    // can be done by FEValues::get_function_values, FEValues::get_function_gradients and FEValues::get_function_divergences.
+
+    std::vector<Tensor<1, dim>>   present_velocity_values    (n_q_points);
```

Comment:
Using `std::vector<Tensor<1, dim>>` rather than `std::vector<Tensor<1, dim> >` will prevent compilation without C++11 (here and elsewhere)